pkg/encryption/path_new.go

```diff
@@ -0,0 +1,290 @@
+			return "", errs.Wrap(err)
+		}
+		if i > 0 {
+			builder.WriteByte('/')
+		}
+		builder.WriteString(unencComponent)
+	}
+	return builder.String(), nil
+}
+
+// StoreDeriveContentKey returns the content key for the passed in path by looking up
+// the appropriate base key from the store and bucket and deriving the rest.
+func StoreDeriveContentKey(bucket string, path paths.Unencrypted, store *Store) (key *storj.Key, err error) {
+	key, err = StoreDerivePathKey(bucket, path, store)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+	key, err = DeriveKey(key, "content")
+	return key, errs.Wrap(err)
+}
+
+// StoreDerivePathKey returns the path key for the passed in path by looking up the
+// appropriate base key from the store and bucket and deriving the rest.
+func StoreDerivePathKey(bucket string, path paths.Unencrypted, store *Store) (key *storj.Key, err error) {
+	_, consumed, base := store.LookupUnencrypted(bucket, path)
+	if base == nil {
+		return nil, errs.New("unable to find encryption base for: %s/%q", bucket, path)
+	}
+
+	// If asking for the key at the bucket, do that and return.
+	if !path.Valid() {
+		key, err = DeriveKey(&base.Key, "path:"+bucket)
+		if err != nil {
+			return nil, errs.Wrap(err)
+		}
+		return key, nil
+	}
+
+	remaining, ok := path.Consume(consumed)
+	if !ok {
+		return nil, errs.New("unable to derive path key for: %s/%q", bucket, path)
+	}
+
+	// if we didn't consume any path, we're at the root of the bucket, and so we have
+	// to fold the bucket name into the key.
+	key = &base.Key
+	if !consumed.Valid() {
+		key, err = DeriveKey(key, "path:"+bucket)
+		if err != nil {
+			return nil, errs.Wrap(err)
+		}
+	}
+
+	for iter := remaining.Iterator(); !iter.Done(); {
+		key, err = DeriveKey(key, "path:"+iter.Next())
+		if err != nil {
+			return nil, errs.Wrap(err)
+		}
+	}
+	return key, nil
+}
+
+// storeEncryptPathComponent encrypts a single path component with the provided cipher and key.
+func storeEncryptPathComponent(comp string, cipher storj.Cipher, key *storj.Key) (string, error) {
+	if comp == "" {
+		return "", nil
+	}
+
+	// derive the key for the next path component. this is so that
+	// every encrypted component has a unique nonce.
+	derivedKey, err := DeriveKey(key, "path:"+comp)
+	if err != nil {
+		return "", err
+	}
+
+	// use the derived key to derive the nonce
+	mac := hmac.New(sha512.New, derivedKey[:])
+	_, err = mac.Write([]byte("nonce"))
+	if err != nil {
+		return "", Error.Wrap(err)
+	}
+
+	nonce := new(storj.Nonce)
+	copy(nonce[:], mac.Sum(nil))
+
+	// encrypt the path components with the parent's key and the derived nonce
+	cipherText, err := Encrypt([]byte(comp), cipher, key, nonce)
+	if err != nil {
+		return "", Error.Wrap(err)
+	}
+
+	nonceSize := storj.NonceSize
+	if cipher == storj.AESGCM {
+		nonceSize = AESGCMNonceSize
+	}
+
+	// keep the nonce together with the cipher text
+	return base64.RawURLEncoding.EncodeToString(append(nonce[:nonceSize], cipherText...)), nil
```

Comment:
whats the rule on base58 v.s. base64?

Reply:
I don't know, but this has to be base64 forever.

Reply:
👍